libs/lz4/src/main/java/org/elasticsearch/lz4/LZ4SafeUtils.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+/*
+ * Copyright 2020 Adrien Grand and the lz4-java contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.elasticsearch.lz4;
+
+import net.jpountz.lz4.LZ4Exception;
+import net.jpountz.util.Utils;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Arrays;
+
+import static org.elasticsearch.lz4.LZ4Constants.LAST_LITERALS;
+import static org.elasticsearch.lz4.LZ4Constants.ML_BITS;
+import static org.elasticsearch.lz4.LZ4Constants.ML_MASK;
+import static org.elasticsearch.lz4.LZ4Constants.RUN_MASK;
+
+/**
+ * This file is forked from https://github.com/lz4/lz4-java. In particular, it forks the following file
+ * net.jpountz.lz4.LZ4SafeUtils.
+ *
+ * It modifies the original implementation to use Java9 array mismatch method and varhandle performance
+ * improvements. Comments are included to mark the changes.
+ */
+enum LZ4SafeUtils {
+    ;
+
+    // Added VarHandle
+    private static final VarHandle intPlatformNative = MethodHandles.byteArrayViewVarHandle(int[].class, Utils.NATIVE_BYTE_ORDER);
+    private static final VarHandle longPlatformNative = MethodHandles.byteArrayViewVarHandle(long[].class, Utils.NATIVE_BYTE_ORDER);
+
+    static int hash(byte[] buf, int i) {
+        return LZ4Utils.hash(SafeUtils.readInt(buf, i));
+    }
+
+    static int hash64k(byte[] buf, int i) {
+        return LZ4Utils.hash64k(SafeUtils.readInt(buf, i));
+    }
+
+    static boolean readIntEquals(byte[] buf, int i, int j) {
+        return SafeUtils.readInt(buf, i) == SafeUtils.readInt(buf, j);
+    }
+
+    static void safeIncrementalCopy(byte[] dest, int matchOff, int dOff, int matchLen) {
+        for (int i = 0; i < matchLen; ++i) {
```

Comment:
Can this loop be replaced with `if (matchLen >= 0) System.arraycopy(dest, matchOff, dest, dOff, matchLen)`?